### ISSUE TITLE
train for a static scene

### ISSUE THREAD

**Issue by HyoKong** [CVMI-Lab/SC-GS#22]
Hi, thanks for the great work!

After I train a static scene, there are so many floating points near the camera.
![Screenshot from 2024-03-25 15-01-21](https://github.com/yihua7/SC-GS/assets/32049327/c6ea710d-370e-436e-904c-9f48bd3400c1)

The PSNR of the training and testing views, however, are very high, 32 and 31, respectively. 

Here is my training script: 
`
CUDA_VISIBLE_DEVICES=0 python train_gui.py --source_path data/in2n-data/face/ --model_path outputs/face --deform_type node --node_num 512 --hyper_dim 8 --eval --gt_alpha_mask_as_scene_mask --local_frame --resolution 2 --W 800 --H 800   --is_scene_static --gui
`

Could you please help to figure it out? Is there any issue with my implementation?

Thank you so much in advance!

**Comment by yihua7** [comment]
Hello,

The current setup of the scene is face-forward (oriented directly toward the viewer), which might pose a significant challenge for Gaussian Splatting to generate an accurate representation of a deviated perspective as depicted in your image. I suggest considering an alternative approach to check this by replacing the parameter `--deform_type node` with `--deform_type static`. This adjustment would allow for the scene to be reconstructed using vanilla static Gaussian Splatting. By doing so, you can assess whether there are any issues within the deform code that could be causing unexpected results.

I suppose such face-forward scenes, lacking 360 degree photometric infos, are too challenging for vanilla Gaussian Splatting.

**Comment by yihua7** [comment]
Indeed, Gaussian Splatting remains effective in face-forward scenes during camera interpolation. It's why the PSNRs at the test view are still high. However, its performance is compromised when utilized for extrapolation view synthesis, as shown in your image.

**Comment by HyoKong** [comment]
Hi, yihua, thanks for the explanation. After I change the deform_type to static, there are no floating points anymore. I guess the reason for floating points is the sampling strategy for the node. When training till `opt.iterations_node_sampling` iteration, the 512 nodes are re-sampled from the GS by the farthest sampling strategy. Maybe some 'noise nodes' near the camera are sampled by the farthest sampling strategy and more floating GS are densified near the 'noise nodes'.

I'm not sure if the suspection is correct. Pls feel free to correct me if there is something wrong or not convincing.

 Thank you!

**Comment by yihua7** [comment]
Hi, thank you for your feedback! I recommend using the provided [script ](https://github.com/yihua7/SC-GS/blob/master/data_tools/phone_catch.py) in this repository to mask the background and keep the object you wish to edit (e.g., the human in your scene). By masking multi-view images with a transparent background, you can potentially eliminate any unwanted floaters and achieve more coherent editing specifically on the human target.
During training you can use `--gt_alpha_mask_as_dynamic_mask --gs_with_motion_mask` to decompose the static and dynamic (not necessary dynamic, editable) parts and rendering them both together.
This is the way I made the DEMOs of T&T scenes Family and Horse. Hope this information helps! : )

**Comment by HyoKong** [comment]
I'll have a try. Thank you so much for your advice and guidance!

**Comment by ENGFCC** [comment]
Good afternoon! Congratulations for your great work!
My training data is **a static scene from IN2N dataset**, it writes"**L_reduced is not invertible, use pseudo inverse instead**" when I use Q from keyboard and right click to edit. And then the gui dosen't work. 
I have no idea about this. Could you please help me? Many thanks!

Here is my training script:
CUDA_VISIBLE_DEVICES=0 python train_gui.py --source_path /home/ubuntu/data/fcc/SC-GS/data/person --model_path outputs/person --deform_type node --node_num 512 --hyper_dim 8 --is_scene_static --init_isotropic_gs_with_all_colmap_pcl --eval --gt_alpha_mask_as_dynamic_mask --gs_with_motion_mask --local_frame --resolution 2 --W 800 --H 800 --gui

![Image](https://github.com/user-attachments/assets/b3ac7051-d809-4575-9efb-d7eaafd2fa74)

**Comment by yihua7** [comment]
Hi,
Actually, the print message is a warning that the Laplacian matrix is invertible but it does not influence anything since the variable is not used. However, the redundant code there affects the editing speed and makes you feel GUI doesn't work. Please refer to our latest commit: 0ba45d389d2f9922a504cf7ee6c1f44ca9e2c8ae and README for newest updates. Currently I-N2D scenes can be flexibly edited. Feel free to try it!

**Comment by ENGFCC** [comment]
Thanks for your early reply!
I noticed that your editing results on the IN2N dataset are quite flexible, with control points accurately placed on the human figures. **May I ask if you used masks generated by MiVOS?**

**I utilized MiVOS-generated masks** and made some modifications to the code as follows:
![Image](https://github.com/user-attachments/assets/7f1179d2-3d3a-4fef-b028-a0270099d12d)
![Image](https://github.com/user-attachments/assets/988fbeb9-d718-4cd9-a366-a512d8d9a629)

The GUI trains normally, but when I **click "Init Graph"**, the terminal **output halts at "Set ball query radius to..."**. The subsequent line, **"Initialize Animation Model with %d control nodes," never appears**. Moreover, the **GUI seems to freeze and becomes unresponsive**.

![Image](https://github.com/user-attachments/assets/fff9f8c2-88a3-43fd-9f94-c85157bb5f02)
![Image](https://github.com/user-attachments/assets/f70503a8-72ac-4002-8bfd-cfafbd901f4b)

Could you please help me identify where I might have gone wrong?
Thank you very much!

**Comment by yihua7** [comment]
1. Yes, I used MIVOS to segment the video. 

2. It may be caused by NaN values in [line](https://github.com/CVMI-Lab/SC-GS/blob/4e3e83139e74a401d94c7dc47a4f492fed8065c5/utils/arap_deform.py#L70). I suggest to check is there any NaN values in self.L_opt.

**Comment by ENGFCC** [comment]
> 2. It may be caused by NaN values in [line](https://github.com/CVMI-Lab/SC-GS/blob/4e3e83139e74a401d94c7dc47a4f492fed8065c5/utils/arap_deform.py#L70). I suggest to check is there any NaN values in self.L_opt.

You're amazing, and you're absolutely right, **there is some NaN in self.L_opt**. It appears that the issue arises from the **cal_connectivity_from_points function within the ARAPDeformer class, where the computed weights contain NaN values.**

However, I'm not sure how to resolve the issue. I truly hope I can achieve the same level of proficiency as you soon.
![Image](https://github.com/user-attachments/assets/550841ce-3e23-46b9-b8c7-ae447b471d0b)

**Comment by yihua7** [comment]
Hi,

Typically the stuck of pytorch calculation is caused by NaN values appearing in its linear algebra functions.

Does your code have this line to avoid Inf values? https://github.com/CVMI-Lab/SC-GS/blob/27d6f1e24295dd7a12f1331b225636623a9772ff/utils/deform_utils.py#L97

If not, I recommend to use the latest code for editing.

Or you can directly remove NaN values in weights via `weight = torch.nan_to_num(weight, nan=0.0)`. You can insert it into anywhere before ARAP deformation like here: https://github.com/CVMI-Lab/SC-GS/blob/27d6f1e24295dd7a12f1331b225636623a9772ff/utils/deform_utils.py#L111.

**Comment by ENGFCC** [comment]
Hello, thank you for your outstanding work!
With your updated code, I was able to train and perform drag operations on the IN2N dataset using the GUI.
However, I apologize for my ignorance — I'm not quite sure **how the rotation operation is implemented**, such as making a face turn.
Also, **after clicking sv_kpt**, should I **click start to continue training**?
Thanks for your early reply!

**Comment by yihua7** [comment]
Thank you for your interest!  

- To rotate the selected group of control points, press **'Q'** while dragging with the right mouse button. For a detailed introduction, please refer to the guide: [2-min Editing Guidance](https://github.com/CVMI-Lab/SC-GS?tab=readme-ov-file#2-min-editing-guidance).  

- Clicking **`sv_kpt`** saves the edited results of the currently selected keypoints. To revisit the edits later, click **`ld_kpt`** to load them in a loop order. You can always resume training at any time.

**Comment by ENGFCC** [comment]
Thanks for your early reply. Your drag interaction works very well. 
I watched the guide 2-min Editing Guidance many times, and I press 'Q' while dragging with the right mouse button, but it couldn't rotate, just translation.

**Comment by yihua7** [comment]
Hi,
Sorry for the bug. I have updated the code and now 'Q' and 'R' both support rotation. Please try the latest code.
